sklearn/linear_model/sag_fast.pyx

```diff
@@ -0,0 +1,325 @@
+# cython: cdivision=True
+#
+# Authors: Danny Sullivan <dbsullivan23@gmail.com>
+#
+# Licence: BSD 3 clause
+
+import numpy as np
+cimport numpy as np
+import warnings
+from libc.math cimport fabs, exp, log
+from libc.time cimport time, time_t
+
+from ..utils.seq_dataset cimport SequentialDataset
+from .sgd_fast cimport Log, SquaredLoss, LossFunction
+
+cdef extern from "sgd_fast_helpers.h":
+    bint skl_isfinite(double) nogil
+
+# This sparse implementation is taken from section 4.3 of "Minimizing Finite
+# Sums with the Stochastic Average Gradient" by
+# Mark Schmidt, Nicolas Le Roux, Francis Bach. 2013. <hal-00860051>
+#
+# https://hal.inria.fr/hal-00860051/PDF/sag_journal.pdf
+
+
+cdef inline double fmax(double x, double y) nogil:
+    if x > y:
+        return x
+    return y
+
+
+def sag(SequentialDataset dataset,
+        np.ndarray[double, ndim=1, mode='c'] weights_array,
+        double intercept,
+        int n_samples,
+        int n_features,
+        double tol,
+        int max_iter,
+        LossFunction loss,
+        double eta,
+        double alpha,
+        np.ndarray[double, ndim=1, mode='c'] sum_gradient_init,
+        np.ndarray[double, ndim=1, mode='c'] gradient_memory_init,
+        np.ndarray[bint, ndim=1, mode='c'] seen_init,
+        int num_seen,
+        double weight_pos,
+        double weight_neg,
+        bint fit_intercept,
+        double intercept_sum_gradient,
+        double intercept_decay,
+        bint verbose):
+
+    # true if the weights or intercept are NaN or infinity
+    cdef bint infinity = False
+    # the pointer to the coef_ or weights
+    cdef double* weights = <double * >weights_array.data
+    # the data pointer for X, the training set
+    cdef double *x_data_ptr
+    # the index pointer for the column of the data
+    cdef int *x_ind_ptr
+    # the label for the sample
+    cdef double y
+    # the sample weight
+    cdef double sample_weight
+    # the number of non-zero features for this sample
+    cdef int xnnz
+    # helper variable for indexes
+    cdef int idx
+    # the total number of interations through the data
+    cdef int total_iter = 0
```

Comment:
This variable name is misleading. It should better be named `n_updates` and the above inline comment:

```
# the total number of individual weight updates applied to the model
```